nixos/modules/services/networking/prosody.nix

```diff
@@ -1,885 +1,659 @@
+      mapAttrsToList (k: v: "The option `services.prosody.virtualHosts.${k}.enabled' has been renamed to `services.prosody.virtualHosts.${k}.settings.enabled'.") (filterAttrs (_: v: v.enabled != null) cfg.virtualHosts) ++
+      mapAttrsToList (k: v: "The option `services.prosody.virtualHosts.${k}.ssl' has been renamed to `services.prosody.virtualHosts.${k}.settings.ssl'.") (filterAttrs (_: v: v.ssl != null) cfg.virtualHosts)
+    ;
+
+    services.prosody.settings = {
+      log = mkDefault { warn = "*syslog"; error = "*syslog"; };
+      data_path = mkForce cfg.dataDir;
+      network_backend = "event";
+      pidfile = "/run/prosody/prosody.pid";
+
+      authentication = mkDefault "internal_hashed";
+      reload_modules = mkIf (acmeHosts != []) [ "tls" ];
+
+      modules_enabled = [
+        # required for compliance with https://compliance.conversations.im/about/
+        "dialback"
+        "disco"
+        "roster"
+        "saslauth"
+        "tls"
+
+        # not essential, but recommended
+        "blocklist"
+        "bookmarks"
+        "carbons"
+        "cloud_notify"
+        "csi"
+        "pep"
+        "private"
+        "vcard_legacy"
+
+        # nice to have
+        "mam"
+        "ping"
+        "register"
+        "smacks"
+        "time"
+        "uptime"
+        "version"
+
+        # admin interfaces
+        "admin_adhoc"
+        "http_files"
+        "proxy65"
+      ] ++ cfg.package.communityModules
+        ++ mapAttrsToList (k: _: k) (filterAttrs (_: v: v == true) cfg.modules);
+
+      modules_disabled = mapAttrsToList (k: _: k) (filterAttrs (_: v: v == false) cfg.modules);
+
+      disco_items =
+        mapAttrsToList (k: v: [ k "${k} HTTP upload endpoint" ]) (filterAttrs (k: v: v.module == "http_upload") cfg.components) ++
+        mapAttrsToList (k: v: [ k "${k} MUC endpoint" ]) (filterAttrs (k: v: v.module == "muc") cfg.components) #++
+      ;
+
+      # mod_tls configuration
+      c2s_require_encryption = mkDefault true;
+      s2s_require_encryption = mkDefault true;
+
+      # upstream defaults - useful for `services.prosody.openFirewall` logic
+      c2s_ports = mkDefault [ 5222 ];
+      s2s_ports = mkDefault [ 5269 ];
+      https_ports = mkDefault [ 5281 ];
+      # TODO:
+      # proxy65_ports = [ 5000 ];
     };
 
-    users.groups.prosody = mkIf (cfg.group == "prosody") {
-      gid = config.ids.gids.prosody;
-    };
+    environment.systemPackages = [ cfg.package ];
+    environment.etc."prosody/prosody.cfg.lua".text =
+      let
+        toFormat = attrs: concatStringsSep "\n" (mapAttrsToList (k: v: ''${k} = ${toStr v}'') (filterAttrs (_: v: v != null) attrs));
+
+        toStr = v:
+          if builtins.isString v then
+            # prosody will directly pass environment variables into its configuration file which have `ENV_` as a prefix
+            if hasPrefix "ENV_" v then v else ''"${toString v}"''
```

Comment:
Can you give an example for this and write it really prominent into the configuration? This would be immensely useful for STUN/TURN.

Reply:
ok i pushed a separate commit so you could review more easily... please feel free to suggest any wording changes which will make this more prominent 🙇‍♂️